### PR TITLE
PKIX cert tooling: tdns-cli cert (minimal internal CA) + cert init + PKIX loose ends

### PR DESCRIPTION
**Pull request by johanix**
Implements docs/2026-07-21-pkix-cert-tooling-design.md (merged to main via #315). **Stacked on feature/xot (#314)** — the PKI reuses `SPKISHA256`/`NewTlsaRR` and the XoT test harness from that branch, so this PR is based on it and should merge after it.

## What this adds

**`tdns-cli cert` — a deliberately minimal internal PKI** (Part 2 / LE-1, the one real blocker for PKIX-based XoT). Pure-stdlib crypto in `v2/pki.go`, thin cobra wrappers in `v2/cli/cert_cmds.go`:

- `cert ca` — self-signed root, hard-coded pathlen 0 (leaves only, never a sub-CA), random 128-bit serials (no serial-file/DB state), ed25519 default (ecdsa-p256/rsa2048 available)
- `cert leaf` — CA-signed end-entity cert; `--emit-pin` / `--emit-tlsa <owner> --tlsa-port` (default 853) print the pin and the TLSA 3-1-1 record, so one issuance feeds all three XoT auth modes
- `cert csr` / `cert sign` — split provisioning for cross-org mutual TLS; the private key never leaves the requesting host
- `cert pin` / `cert show` — read-only helpers
- **`cert init`** — the tier-1 one-shot for third-party users: reads the local tdns-auth config, creates the CA if absent (reuses otherwise), derives SANs from the config's listen addresses + hostname, writes cert/key to the exact `certfile`/`keyfile` paths the config names, and prints ready-to-paste secondary config for pkix/pin/dane. Idempotent; `--force` to replace.

State is **plain PEM files** (CA home `/etc/tdns/ca/`, dir 0700, keys 0600, no overwrite without `--force`) plus an append-only human-readable `issued.log` — an audit trail, deliberately not a cert database. No CRL/OCSP/renewal; scope stated in `--help`. The `cert` subtree is top-level in tdns-cli (per the design doc; could move under `util` if preferred) and skips config/API init — it works with no daemon present.

**PKIX loose ends (Part 1):**
- LE-2: startup note for a single CA-signed cert in certfile + "leaf first, bundle intermediates" docs. Info rather than the doc's suggested Warn: that shape is exactly what `cert init` produces (two-tier PKI, root trusted directly), so a warning would cry wolf on the blessed flow.
- LE-3: agent-parity test driving the full production `FetchFromUpstream` path (hard flip + snapshot publish) over dot+pkix.
- LE-4 (own commit, easy to drop): opt-in `dnsengine.downstream-names` SAN allowlist for shared-CA mTLS; empty = unchanged chain-only behavior.
- LE-5: not implemented (non-goal per the doc).
- LE-6: dog prints an IP-SAN hint when `+cafile` targets an IP-literal server.
- LE-7: intermediate-chain test (hand-built root→intermediate→leaf; transfer succeeds with the chain presented, fails without the intermediate). SAN-mismatch was already covered on feature/xot.
- Ops guide: new tiered "Provisioning certificates" section — casual testers need none of this (mwe self-signed + pin/dane still the first path), `cert init` for one host, primitives for fleets.

## Testing

- PKI unit tests: CA constraints + self-verify, leaf chain/EKU/name verification, leaf-cannot-sign, CSR round-trip incl. tamper refusal, all three algorithms.
- Closing-the-loop integration: a minted CA + leaf satisfies all three XoT client auth modes (pkix/pin/dane) against the real transfer path.
- `go test -race` green across v2, v2/cli, v2/cache, v2/core, v2/edns0; all six binaries build.
- Live smoke: full `ca`/`leaf`/`csr`/`sign`/`pin`/`show` walkthrough (0600 keys, overwrite refusal, issued.log); `cert init` against an MWE config → daemon restart → dog `+cafile` PKIX-verified AXFR with the minted CA; re-run reuses the CA and refuses to clobber the leaf.

## Update (9144d15): SPKI-preserving upgrade path + provisioning guide

- `cert csr --key` (re-certify an existing key; SPKI unchanged so pins/TLSA stay valid) and `--from-cert` (copy CN+SANs); `ParsePrivateKeyPEM` accepts legacy openssl EC/RSA PEM keys.
- New **guide/cert-provisioning.md** (indexed in guide/README.md): worked examples for upgrading local and remote self-signed certs to CA-signed, building the ca-file for each kind of cert, renewal/rotation, and the no-revocation scope.
- Live-checked end-to-end: MWE self-signed cert upgraded in place (issuer flipped, pin unchanged), then a two-daemon pkix XoT pull + dog `+cafile` query against the CA.

## Summary by CodeRabbit

* **New Features**
  * Added offline `cert` CLI tooling for CA/leaf workflows, CSR generation/signing, and producing pins/TLSA records.
  * Added `cert init` for one-step local provisioning with automatic SAN detection.
  * Added CA-mode downstream client certificate DNS-name allowlists for stricter mTLS identity checks.
* **Improvements**
  * More resilient encrypted transport TLS cert bootstrapping and clearer certificate-chain messaging.
  * Improved guidance when IP-SAN PKIX verification fails.
* **Tests**
  * Added end-to-end PKI/TLSA/pin and PKIX edge-case coverage, plus CLI flag-binding regression tests.

### CODE REVIEW

**Comment by johanix** [comment]
@coderabbitai review

**Comment by johanix** [comment]
@coderabbitai full review

File: v2/do53.go [hunk 1]
```diff
@@ -127,39 +127,55 @@ func DnsEngine(ctx context.Context, conf *Config) error {
...
+			certReason = fmt.Sprintf("reading keyfile %s: %v", keyFile, err)
+		} else {
+			conf.Internal.CertData = string(certPEM)
+			conf.Internal.KeyData = string(keyPEM)
+
+			if cert, err = tls.LoadX509KeyPair(certFile, keyFile); err != nil {
+				lgDns.Error("DnsEngine: failed to load certificate, not starting DoT/DoH/DoQ service", "err", err)
+				certKey = false
+				certReason = fmt.Sprintf("loading certfile/keyfile: %v", err)
+			}
```
[the remark this answers is not in the document]
Comment (johanix) [reply v2/do53.go:181]: Fixed in `e4e848f`: `tls.LoadX509KeyPair(certFile, keyFile)` → `tls.X509KeyPair(certPEM, keyPEM)`, so the cert is parsed from the bytes already read (and stored in `CertData`/`KeyData`) rather than a second disk read. One read, no drift.

While here, also addressed the **outside-diff** comment on this same file (cert/key access failures), in `5e5e075`: `os.Stat` now catches *every* access failure (not only `os.IsNotExist`), and an unreadable / permission-denied / directory cert degrades gracefully (`certKey=false` + `certReason` → `SetTransportCertError`) instead of a fatal `return` — so `config status` reports it and Do53 keeps running, consistent with the missing-file and parse-failure paths.

Resolving.

File: v2/cli/cert_cmds_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		t.Fatalf("cert ca --validity default = %q, want 3650", caFlag.DefValue)
+	}
+	// The CA flag must NOT share the leaf/sign backing variable, or
+	// addCommonLeafFlags rebinds it to 397.
+	if certCAValidity != 3650 {
+		t.Fatalf("certCAValidity = %d after init, want 3650 (shared-variable regression)", certCAValidity)
+	}
+	if certValidity != 397 {
+		t.Fatalf("certValidity = %d after init, want 397 (leaf/sign default)", certValidity)
+	}
```
[the remark this answers is not in the document]
Comment (johanix) [reply v2/cli/cert_cmds_test.go:28]: Fixed in `5e5e075`: `TestCertValidityFlagBindings` now sets the CA `--validity` value and asserts only `certCAValidity` changes (`certValidity` untouched, restored via `t.Cleanup`) — verifying the binding directly rather than the defaults alone. (Thread already resolved.)
